x/delegation/keeper/update_native_restaking_balance.go

```diff
@@ -0,0 +1,131 @@
+		})
+		if err != nil {
+			return err
+		}
+		ctx.Logger().Info("UpdateNativeRestakingBalance slash from withdrawable amount", "stakerID", stakerID, "assetID", assetID, "slashFromWithdrawable", slashFromWithdrawable, "remainAmount", remainAmount)
+
+		// slash from pending undelegations
+		if remainAmount.IsPositive() {
+			opFunc := func(undelegationKey string, undelegation *delegationtypes.UndelegationRecord) (bool, error) {
+				// slash from the single undelegation
+				slashAmount := remainAmount
+				remainAmount = slashAmount.Sub(undelegation.ActualCompletedAmount)
+				if remainAmount.IsPositive() {
+					slashAmount = undelegation.ActualCompletedAmount
+				}
+				undelegation.ActualCompletedAmount = undelegation.ActualCompletedAmount.Sub(slashAmount)
+				err = k.assetsKeeper.UpdateStakerAssetState(ctx, stakerID, assetID, types.DeltaStakerSingleAsset{
+					TotalDepositAmount: slashAmount.Neg(),
+				})
+				if err != nil {
+					return true, err
+				}
+				ctx.Logger().Info("UpdateNativeRestakingBalance slash from undelegation", "stakerID", stakerID, "assetID", assetID, "operator", undelegation.OperatorAddr, "undelegationKey", undelegationKey, "slashAmount", slashAmount, "remainAmount", remainAmount)
+				if !remainAmount.IsPositive() {
+					// return ture to break the iteration if there isn't remaining amount to be slashed
+					return true, nil
+				}
+				return false, nil
+			}
+			err = k.IterateUndelegationsByStakerAndAsset(ctx, stakerID, assetID, true, opFunc)
+			if err != nil {
+				return err
+			}
+		}
+
+		// slash from the delegated share
+		// the delegated share will be proportionally decreased from all operators to
+		// whom the related staker has already delegated
+		if remainAmount.IsPositive() {
+			// calculate the slash proportion
+			totalDelegatedAmount, err := k.TotalDelegatedAmountForStakerAsset(ctx, stakerID, assetID)
+			if err != nil {
+				return err
+			}
+			if !totalDelegatedAmount.IsZero() {
+				slashProportion := sdkmath.LegacyNewDecFromBigInt(remainAmount.BigInt()).Quo(sdkmath.LegacyNewDecFromBigInt(totalDelegatedAmount.BigInt()))
+				if slashProportion.GT(sdkmath.LegacyNewDec(1)) {
+					slashProportion = sdkmath.LegacyNewDec(1)
+				}
+				opFunc := func(keys *delegationtypes.SingleDelegationInfoReq, delegationAmount *delegationtypes.DelegationAmounts) (bool, error) {
+					opAccAddr, err := sdk.AccAddressFromBech32(keys.OperatorAddr)
+					if err != nil {
+						return true, err
+					}
+					slashShare := delegationAmount.UndelegatableShare.Mul(slashProportion)
+					actualSlashAmount, err := k.RemoveShare(ctx, false, opAccAddr, stakerID, assetID, slashShare)
+					if err != nil {
+						return true, err
+					}
+					err = k.assetsKeeper.UpdateStakerAssetState(ctx, stakerID, assetID, types.DeltaStakerSingleAsset{
+						TotalDepositAmount: actualSlashAmount.Neg(),
+					})
+					if err != nil {
+						return true, err
+					}
+					remainAmount = remainAmount.Sub(actualSlashAmount)
```

Comment:
it is better to use _pendingSlashAmount_ to replace remainAmount

Reply:
fixed